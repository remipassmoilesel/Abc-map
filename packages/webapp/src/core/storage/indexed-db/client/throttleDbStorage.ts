/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import throttle from 'lodash/throttle';

/**
 * This function allow to limit storage call.
 *
 * With a 5s value, it will ensure that storageCb will be called once then no more the 5 next seconds.
 *
 * @param storageCb
 * @param waitBetweenCallsMs
 */
export function throttleDbStorage(storageCb: (...args: any[]) => void, waitBetweenCallsMs = 5_000) {
  return throttle(storageCb, waitBetweenCallsMs, { leading: true, trailing: true });
}
