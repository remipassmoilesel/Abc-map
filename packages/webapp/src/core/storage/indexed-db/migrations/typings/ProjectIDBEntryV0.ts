/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcView } from '@abc-map/shared';

export interface ProjectIDBEntryV0 {
  version?: number;
  // Metadata: id, version, etc ...
  metadata: IDBProjectMetadataV0;
  // Main view for edition
  view: AbcView;
  layerIds: string[];
  // Layouts for static exports
  layoutIds: string[];
  // Views publicly accessible
  sharedViews: {
    // If true, map is fullscreen
    fullscreen: boolean;
    // If not fullscreen, these dimensions are used
    mapDimensions: { width: number; height: number };
    // List of views
    viewIds: string[];
  };
}

export interface IDBProjectMetadataV0 {
  id: string;
  version: string;
  name: string;
  public: boolean;
  // This field has not been removed in a previous migration
  containsCredentials?: boolean;
}
