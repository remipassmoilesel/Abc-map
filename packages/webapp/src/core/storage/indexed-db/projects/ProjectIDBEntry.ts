/**
 * Copyright © 2022 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcProjectMetadata, AbcView } from '@abc-map/shared';

export const CURRENT_VERSION = 3;

export interface ProjectIDBEntry {
  version: number;
  // Metadata: id, version, etc ...
  metadata: AbcProjectMetadata;
  // Main view for edition
  view: AbcView;
  layerIds: string[];
  // Layouts for PDF exports
  layouts: {
    layoutIds: string[];
    abcMapAttributionsEnabled: boolean;
  };
  // Views publicly accessible
  sharedViews: {
    // If true, map is fullscreen
    fullscreen: boolean;
    // If not fullscreen, these dimensions are used
    mapDimensions: { width: number; height: number };
    // List of views
    viewIds: string[];
  };
}
