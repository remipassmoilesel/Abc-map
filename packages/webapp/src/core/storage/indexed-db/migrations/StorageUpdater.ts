/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { Logger } from '@abc-map/shared';
import { getMigrations, MigrationsFactory } from './migrations';
import { IndexedDbClient } from '../client/IndexedDbClient';
import { StorageMigrationContext } from './IndexedDbMigration';
import { MigrationIDBStorage } from './MigrationIDBStorage';

export const logger = Logger.get('StorageUpdater.ts', 'debug');

export function disableStorageMigrationLogs() {
  logger.disable();
}

/**
 * This helper migrate Indexed DB data
 */
export class StorageUpdater {
  public static create() {
    return new StorageUpdater(getMigrations);
  }

  constructor(private migrations: MigrationsFactory) {}

  public async update(client: IndexedDbClient): Promise<void> {
    const migrations = this.migrations();
    const migrationStorage = new MigrationIDBStorage(client);
    const migrationsAlreadyDone = await migrationStorage.getAll();

    const context: StorageMigrationContext = { client };

    for (const migration of migrations) {
      const alreadyDone = migrationsAlreadyDone.find((migr) => migr.name === migration.getName());
      if (!alreadyDone) {
        logger.warn(`Executing IndexedDB schema migration: ${migration.getName()}`);
        await migration.process(context);
        await migrationStorage.put({ name: migration.getName(), when: Date.now() });
      }
    }
  }
}
