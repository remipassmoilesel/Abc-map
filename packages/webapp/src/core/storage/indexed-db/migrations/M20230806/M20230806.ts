/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { IndexedDbMigration, StorageMigrationContext } from '../IndexedDbMigration';
import { ProjectIDBEntryV0 } from '../typings/ProjectIDBEntryV0';
import { ObjectStore } from '../../client/ObjectStore';
import { ProjectIDBEntry } from '../../projects/ProjectIDBEntry';
import { GenericReduxIDBEntry } from '../../redux/GenericReduxIDBEntry';
import { AbcLayout, AbcSharedView, Logger } from '@abc-map/shared';
import { FeatureIDBEntry } from '../../features/FeatureIDBEntry';
import { KvPair } from '../../client/IndexedDbClient';
import { TileIDBEntry } from '../../tiles/TileIDBEntry';
import { LayerIDBEntry } from '../../layers/LayerIDBEntry';
import { FeatureIDBEntryV0 } from '../typings/FeatureIDBEntryV0';

const NEXT = 2;

const logger = Logger.get('M20230816.ts');

/**
 * This migration:
 * - updates layout structure
 * - add ids on all entities
 */
export class M20230806 implements IndexedDbMigration {
  public getName(): string {
    return 'M20230806';
  }

  public async process(context: StorageMigrationContext): Promise<void> {
    const { client } = context;

    // Projects
    const projects = await client.getAll<ProjectIDBEntryV0>(ObjectStore.Projects);
    const migratedProjects = projects
      .filter((entry) => (entry.value.version ?? 0) < NEXT)
      .map<ProjectIDBEntry>((entry) => {
        const result: ProjectIDBEntry = {
          version: NEXT,
          metadata: {
            ...entry.value.metadata,
            // We must set appropriate version, otherwise schema can be migrated another time at loading
            version: '1.3.0',
          },
          view: entry.value.view,
          layouts: {
            layoutIds: entry.value.layoutIds,
            abcMapAttributionsEnabled: true,
          },
          sharedViews: entry.value.sharedViews,
          layerIds: entry.value.layerIds,
        };

        return result;
      });

    if (migratedProjects.length) {
      await client.putAll<ProjectIDBEntry>(
        ObjectStore.Projects,
        migratedProjects.map((entry) => ({ key: entry.metadata.id, value: entry }))
      );
    }

    // Layouts
    const layouts = await client.getAll<AbcLayout | GenericReduxIDBEntry<AbcLayout>>(ObjectStore.Layouts);
    const migratedLayouts = layouts
      .filter((entry) => ((entry.value as GenericReduxIDBEntry<AbcLayout>).version ?? 0) < NEXT)
      .map<GenericReduxIDBEntry<AbcLayout>>((entry) => {
        return {
          id: entry.value.id,
          version: NEXT,
          entry: entry.value as AbcLayout,
        };
      });

    if (migratedLayouts.length) {
      await client.putAll<GenericReduxIDBEntry<AbcLayout>>(
        ObjectStore.Layouts,
        migratedLayouts.map((entry) => ({ key: entry.id, value: entry }))
      );
    }

    // Layers
    const layers = await client.getAll<LayerIDBEntry>(ObjectStore.Layers);
    const migratedLayers = layers
      .filter((entry) => (entry.value.version ?? 0) < NEXT)
      .map<LayerIDBEntry>((entry) => {
        return {
          version: NEXT,
          layer: entry.value.layer,
        };
      });

    if (migratedLayers.length) {
      await client.putAll<LayerIDBEntry>(
        ObjectStore.Layers,
        migratedLayers.map((entry) => ({ key: entry.layer.metadata.id, value: entry }))
      );
    }

    // Shared views
    const sharedViews = await client.getAll<AbcSharedView | GenericReduxIDBEntry<AbcSharedView>>(ObjectStore.SharedViews);
    const migratedSharedViews = sharedViews
      .filter((entry) => ((entry.value as GenericReduxIDBEntry<AbcSharedView>).version ?? 0) < NEXT)
      .map<GenericReduxIDBEntry<AbcSharedView>>((entry) => {
        return {
          id: entry.value.id,
          version: NEXT,
          entry: entry.value as AbcSharedView,
        };
      });

    if (migratedSharedViews.length) {
      await client.putAll<GenericReduxIDBEntry<AbcSharedView>>(
        ObjectStore.SharedViews,
        migratedSharedViews.map((entry) => ({ key: entry.id, value: entry }))
      );
    }

    // Features
    const features = await client.getAll<FeatureIDBEntryV0>(ObjectStore.Features);
    const migratedFeatures = features
      .filter((entry) => (entry.value.version ?? 0) < NEXT)
      .map<FeatureIDBEntry>((entry) => {
        return {
          version: NEXT,
          layerId: entry.value.layerId,
          feature: entry.value.feature,
        };
      });

    if (migratedFeatures.length) {
      await client.putAll<FeatureIDBEntry>(
        ObjectStore.Features,
        migratedFeatures
          .map<KvPair | null>((feature) => {
            if (!feature.feature.id) {
              logger.warn('Warning: this feature does not have an id, it will not be saved');
              return null;
            }

            return { key: feature.feature.id, value: feature };
          })
          .filter((u): u is KvPair<FeatureIDBEntry> => !!u)
      );
    }

    // Tiles
    const tiles = await client.getAll<Blob>(ObjectStore.Tiles);
    const migratedTiles = tiles
      .filter((entry) => ((entry.value as any).version ?? 0) < NEXT)
      .map<TileIDBEntry | null>((entry) => {
        if (typeof entry.key !== 'string') {
          return null;
        }

        return {
          version: NEXT,
          url: entry.key,
          image: entry.value,
        };
      })
      .filter((v): v is TileIDBEntry => !!v);

    if (migratedTiles.length) {
      await client.putAll<TileIDBEntry>(
        ObjectStore.Tiles,
        migratedTiles.map<KvPair<TileIDBEntry>>((tile) => {
          return { key: tile.url, value: tile };
        })
      );
    }
  }
}
