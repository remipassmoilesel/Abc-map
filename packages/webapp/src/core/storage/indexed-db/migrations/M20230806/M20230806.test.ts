/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { ProjectIDBEntryV0 } from '../typings/ProjectIDBEntryV0';
import { ObjectStore } from '../../client/ObjectStore';
import { ProjectIDBEntry } from '../../projects/ProjectIDBEntry';
import { IndexedDbClient } from '../../client/IndexedDbClient';
import { nanoid } from 'nanoid';
import { TestHelper } from '../../../../utils/test/TestHelper';
import { M20230806 } from './M20230806';
import { GenericReduxIDBEntry } from '../../redux/GenericReduxIDBEntry';
import { AbcLayout, AbcSharedView } from '@abc-map/shared';
import { FeatureIDBEntry } from '../../features/FeatureIDBEntry';
import { FeatureIDBEntryV0 } from '../typings/FeatureIDBEntryV0';
import { TileIDBEntry } from '../../tiles/TileIDBEntry';
import { TileIDBEntryV0 } from '../typings/TileIDBEntryV0';
import { LayerIDBEntry } from '../../layers/LayerIDBEntry';
import { LayerIDBEntryV0 } from '../typings/LayerIDBEntryV0';
import { disableStorageMigrationLogs } from '../StorageUpdater';

disableStorageMigrationLogs();

describe('M20230806', () => {
  let dbName: string;
  let client: IndexedDbClient;
  let migration: M20230806;

  beforeEach(async () => {
    client = new IndexedDbClient();
    dbName = nanoid();

    await client.connect(dbName, 1, (db) => {
      db.createObjectStore(ObjectStore.Projects);
      db.createObjectStore(ObjectStore.Layouts);
      db.createObjectStore(ObjectStore.Layers);
      db.createObjectStore(ObjectStore.SharedViews);
      db.createObjectStore(ObjectStore.Features);
      db.createObjectStore(ObjectStore.Tiles);
    });

    migration = new M20230806();
  });

  afterEach(() => {
    client.disconnect();
  });

  describe('Projects', () => {
    it('should do nothing', async () => {
      // Prepare
      const template = TestHelper.sampleProjectManifest();
      const docs: ProjectIDBEntry[] = [
        {
          version: 99,
          metadata: {
            ...template.metadata,
            id: 'test-project-1',
            name: 'Test project 1',
          },
          view: template.view,
          layerIds: ['layer-id-1', 'layer-id-2', 'layer-id-3'],
          layouts: {
            layoutIds: ['layout-id-1', 'layout-id-2', 'layout-id-3'],
            abcMapAttributionsEnabled: true,
          },
          sharedViews: {
            fullscreen: template.sharedViews.fullscreen,
            mapDimensions: template.sharedViews.mapDimensions,
            viewIds: ['shared-view-id-1', 'shared-view-id-2', 'shared-view-id-3'],
          },
        },
      ];

      await client.putAll(
        ObjectStore.Projects,
        docs.map((entry) => ({ key: entry.metadata.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<ProjectIDBEntry>(ObjectStore.Projects);
      expect(entries).toMatchSnapshot();
    });

    it('should migrate', async () => {
      // Prepare
      const template = TestHelper.sampleProjectManifest();
      const docs: ProjectIDBEntryV0[] = [
        {
          metadata: {
            ...template.metadata,
            id: 'test-project-1',
            name: 'Test project 1',
          },
          view: template.view,
          layerIds: ['layer-id-1', 'layer-id-2', 'layer-id-3'],
          layoutIds: ['layout-id-1', 'layout-id-2', 'layout-id-3'],
          sharedViews: {
            fullscreen: template.sharedViews.fullscreen,
            mapDimensions: template.sharedViews.mapDimensions,
            viewIds: ['shared-view-id-1', 'shared-view-id-2', 'shared-view-id-3'],
          },
        },
      ];

      await client.putAll(
        ObjectStore.Projects,
        docs.map((entry) => ({ key: entry.metadata.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<ProjectIDBEntry>(ObjectStore.Projects);
      expect(entries).toMatchSnapshot();
    });
  });

  describe('Layouts', () => {
    it('should do nothing', async () => {
      // Prepare
      const docs: GenericReduxIDBEntry<AbcLayout>[] = [
        {
          version: 99,
          id: 'test-layout-1',
          entry: {
            ...TestHelper.sampleLayout(),
            id: 'test-layout-1',
            textFrames: [{ ...TestHelper.sampleTextFrame(), id: 'test-text-frame-1' }],
          },
        },
      ];

      await client.putAll(
        ObjectStore.Layouts,
        docs.map((entry) => ({ key: entry.id, value: entry.entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<GenericReduxIDBEntry<AbcLayout>>(ObjectStore.Layouts);
      expect(entries).toMatchSnapshot();
    });

    it('should migrate', async () => {
      // Prepare
      const docs: AbcLayout[] = [
        {
          ...TestHelper.sampleLayout(),
          id: 'test-layout-1',
          textFrames: [{ ...TestHelper.sampleTextFrame(), id: 'test-text-frame-1' }],
        },
      ];

      await client.putAll(
        ObjectStore.Layouts,
        docs.map((entry) => ({ key: entry.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<GenericReduxIDBEntry<AbcLayout>>(ObjectStore.Layouts);
      expect(entries).toMatchSnapshot();
    });
  });

  describe('Layers', () => {
    it('should do nothing', async () => {
      // Prepare
      const docs: LayerIDBEntry[] = [
        {
          version: 99,
          layer: {
            ...TestHelper.sampleOsmLayer(),
            metadata: {
              ...TestHelper.sampleOsmLayer().metadata,
              id: 'test-layer-1',
            },
          },
        },
      ];

      await client.putAll(
        ObjectStore.Layers,
        docs.map((entry) => ({ key: entry.layer.metadata.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<LayerIDBEntry>(ObjectStore.Layers);
      expect(entries).toMatchSnapshot();
    });

    it('should migrate', async () => {
      // Prepare
      const docs: LayerIDBEntryV0[] = [
        {
          layer: {
            ...TestHelper.sampleOsmLayer(),
            metadata: {
              ...TestHelper.sampleOsmLayer().metadata,
              id: 'test-layer-1',
            },
          },
        },
      ];

      await client.putAll(
        ObjectStore.Layers,
        docs.map((entry) => ({ key: entry.layer.metadata.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<LayerIDBEntry>(ObjectStore.Layers);
      expect(entries).toMatchSnapshot();
    });
  });

  describe('SharedViews', () => {
    it('should do nothing', async () => {
      // Prepare
      const docs: GenericReduxIDBEntry<AbcSharedView>[] = [
        {
          version: 99,
          id: 'test-shared-view-1',
          entry: {
            ...TestHelper.sampleSharedView(),
            id: 'test-shared-view-1',
            textFrames: [{ ...TestHelper.sampleTextFrame(), id: 'test-text-frame-1' }],
          },
        },
      ];

      await client.putAll(
        ObjectStore.SharedViews,
        docs.map((entry) => ({ key: entry.id, value: entry.entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<GenericReduxIDBEntry<AbcSharedView>>(ObjectStore.SharedViews);
      expect(entries).toMatchSnapshot();
    });

    it('should migrate', async () => {
      // Prepare
      const docs: AbcSharedView[] = [
        {
          ...TestHelper.sampleSharedView(),
          id: 'test-shared-view-1',
          textFrames: [{ ...TestHelper.sampleTextFrame(), id: 'test-text-frame-1' }],
        },
      ];

      await client.putAll(
        ObjectStore.SharedViews,
        docs.map((entry) => ({ key: entry.id, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<GenericReduxIDBEntry<AbcSharedView>>(ObjectStore.SharedViews);
      expect(entries).toMatchSnapshot();
    });
  });

  describe('Features', () => {
    it('should do nothing', async () => {
      // Prepare
      const docs: FeatureIDBEntry[] = [
        {
          version: 99,
          layerId: 'test-layer-id-1',
          feature: {
            ...TestHelper.sampleGeojsonFeature(),
            id: 'sample-feature-id',
          },
        },
      ];

      await client.putAll(
        ObjectStore.Features,
        docs.map((entry) => ({ key: entry.feature.id as string, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<FeatureIDBEntry>(ObjectStore.Features);
      expect(entries).toMatchSnapshot();
    });

    it('should migrate', async () => {
      // Prepare
      const docs: FeatureIDBEntryV0[] = [
        {
          layerId: 'test-layer-id-1',
          feature: {
            ...TestHelper.sampleGeojsonFeature(),
            id: 'sample-feature-id',
          },
        },
      ];

      await client.putAll(
        ObjectStore.Features,
        docs.map((entry) => ({ key: entry.feature.id as string, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<FeatureIDBEntry>(ObjectStore.Features);
      expect(entries).toMatchSnapshot();
    });
  });

  describe('Tiles', () => {
    it('should do nothing', async () => {
      // Prepare
      const docs: TileIDBEntry[] = [
        {
          version: 99,
          url: 'http://1',
          image: 'blob data 1' as unknown as Blob,
        },
      ];

      await client.putAll(
        ObjectStore.Tiles,
        docs.map((entry) => ({ key: entry.url, value: entry }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<FeatureIDBEntry>(ObjectStore.Tiles);
      expect(entries).toMatchSnapshot();
    });

    it('should migrate', async () => {
      // Prepare
      const docs: TileIDBEntryV0[] = [
        {
          url: 'http://1',
          image: 'blob data 1' as unknown as Blob,
        },
      ];

      await client.putAll(
        ObjectStore.Tiles,
        docs.map((entry) => ({ key: entry.url, value: entry.image }))
      );

      // Act
      await migration.process({ client });

      // Assert
      const entries = await client.getAll<FeatureIDBEntry>(ObjectStore.Tiles);
      expect(entries).toMatchSnapshot();
    });
  });
});
