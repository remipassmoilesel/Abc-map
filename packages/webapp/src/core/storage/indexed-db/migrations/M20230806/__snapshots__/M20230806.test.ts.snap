// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M20230806 Features should do nothing 1`] = `
Array [
  Object {
    "key": "sample-feature-id",
    "value": Object {
      "feature": Object {
        "bbox": Array [
          1,
          2,
          3,
          4,
        ],
        "geometry": Object {
          "coordinates": Array [
            1,
            5,
          ],
          "type": "Point",
        },
        "id": "sample-feature-id",
        "properties": Object {
          "val": "var",
        },
        "type": "Feature",
      },
      "layerId": "test-layer-id-1",
      "version": 99,
    },
  },
]
`;

exports[`M20230806 Features should migrate 1`] = `
Array [
  Object {
    "key": "sample-feature-id",
    "value": Object {
      "feature": Object {
        "bbox": Array [
          1,
          2,
          3,
          4,
        ],
        "geometry": Object {
          "coordinates": Array [
            1,
            5,
          ],
          "type": "Point",
        },
        "id": "sample-feature-id",
        "properties": Object {
          "val": "var",
        },
        "type": "Feature",
      },
      "layerId": "test-layer-id-1",
      "version": 2,
    },
  },
]
`;

exports[`M20230806 Layers should do nothing 1`] = `
Array [
  Object {
    "key": "test-layer-1",
    "value": Object {
      "layer": Object {
        "metadata": Object {
          "active": false,
          "id": "test-layer-1",
          "model": "OSM",
          "name": "OpenStreetMap",
          "opacity": 1,
          "type": "Predefined",
          "visible": true,
        },
        "type": "Predefined",
      },
      "version": 99,
    },
  },
]
`;

exports[`M20230806 Layers should migrate 1`] = `
Array [
  Object {
    "key": "test-layer-1",
    "value": Object {
      "layer": Object {
        "metadata": Object {
          "active": false,
          "id": "test-layer-1",
          "model": "OSM",
          "name": "OpenStreetMap",
          "opacity": 1,
          "type": "Predefined",
          "visible": true,
        },
        "type": "Predefined",
      },
      "version": 2,
    },
  },
]
`;

exports[`M20230806 Layouts should do nothing 1`] = `
Array [
  Object {
    "key": "test-layout-1",
    "value": Object {
      "entry": Object {
        "format": Object {
          "height": 297,
          "id": "A4_PORTRAIT",
          "orientation": "portrait",
          "width": 210,
        },
        "id": "test-layout-1",
        "name": "Sample layout",
        "textFrames": Array [
          Object {
            "content": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Hey man, how are you ?",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "id": "test-text-frame-1",
            "position": Object {
              "x": 10,
              "y": 10,
            },
            "size": Object {
              "height": 100,
              "width": 100,
            },
            "style": Object {
              "background": "#FFFFFF",
              "withBorders": true,
              "withShadows": true,
            },
          },
        ],
        "view": Object {
          "center": Array [
            1.5,
            45.4,
          ],
          "projection": Object {
            "name": "EPSG:3857",
          },
          "resolution": 1000,
          "rotation": 0,
        },
      },
      "id": "test-layout-1",
      "version": 2,
    },
  },
]
`;

exports[`M20230806 Layouts should migrate 1`] = `
Array [
  Object {
    "key": "test-layout-1",
    "value": Object {
      "entry": Object {
        "format": Object {
          "height": 297,
          "id": "A4_PORTRAIT",
          "orientation": "portrait",
          "width": 210,
        },
        "id": "test-layout-1",
        "name": "Sample layout",
        "textFrames": Array [
          Object {
            "content": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Hey man, how are you ?",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "id": "test-text-frame-1",
            "position": Object {
              "x": 10,
              "y": 10,
            },
            "size": Object {
              "height": 100,
              "width": 100,
            },
            "style": Object {
              "background": "#FFFFFF",
              "withBorders": true,
              "withShadows": true,
            },
          },
        ],
        "view": Object {
          "center": Array [
            1.5,
            45.4,
          ],
          "projection": Object {
            "name": "EPSG:3857",
          },
          "resolution": 1000,
          "rotation": 0,
        },
      },
      "id": "test-layout-1",
      "version": 2,
    },
  },
]
`;

exports[`M20230806 Projects should do nothing 1`] = `
Array [
  Object {
    "key": "test-project-1",
    "value": Object {
      "layerIds": Array [
        "layer-id-1",
        "layer-id-2",
        "layer-id-3",
      ],
      "layouts": Object {
        "abcMapAttributionsEnabled": true,
        "layoutIds": Array [
          "layout-id-1",
          "layout-id-2",
          "layout-id-3",
        ],
      },
      "metadata": Object {
        "id": "test-project-1",
        "name": "Test project 1",
        "public": false,
        "version": "1.3.0",
      },
      "sharedViews": Object {
        "fullscreen": false,
        "mapDimensions": Object {
          "height": 600,
          "width": 800,
        },
        "viewIds": Array [
          "shared-view-id-1",
          "shared-view-id-2",
          "shared-view-id-3",
        ],
      },
      "version": 99,
      "view": Object {
        "center": Array [
          1,
          2,
        ],
        "projection": Object {
          "name": "EPSG:3857",
        },
        "resolution": 1000,
        "rotation": 0,
      },
    },
  },
]
`;

exports[`M20230806 Projects should migrate 1`] = `
Array [
  Object {
    "key": "test-project-1",
    "value": Object {
      "layerIds": Array [
        "layer-id-1",
        "layer-id-2",
        "layer-id-3",
      ],
      "layouts": Object {
        "abcMapAttributionsEnabled": true,
        "layoutIds": Array [
          "layout-id-1",
          "layout-id-2",
          "layout-id-3",
        ],
      },
      "metadata": Object {
        "id": "test-project-1",
        "name": "Test project 1",
        "public": false,
        "version": "1.3.0",
      },
      "sharedViews": Object {
        "fullscreen": false,
        "mapDimensions": Object {
          "height": 600,
          "width": 800,
        },
        "viewIds": Array [
          "shared-view-id-1",
          "shared-view-id-2",
          "shared-view-id-3",
        ],
      },
      "version": 2,
      "view": Object {
        "center": Array [
          1,
          2,
        ],
        "projection": Object {
          "name": "EPSG:3857",
        },
        "resolution": 1000,
        "rotation": 0,
      },
    },
  },
]
`;

exports[`M20230806 SharedViews should do nothing 1`] = `
Array [
  Object {
    "key": "test-shared-view-1",
    "value": Object {
      "entry": Object {
        "id": "test-shared-view-1",
        "layers": Array [],
        "scale": Object {
          "x": 30,
          "y": 100,
        },
        "textFrames": Array [
          Object {
            "content": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Hey man, how are you ?",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "id": "test-text-frame-1",
            "position": Object {
              "x": 10,
              "y": 10,
            },
            "size": Object {
              "height": 100,
              "width": 100,
            },
            "style": Object {
              "background": "#FFFFFF",
              "withBorders": true,
              "withShadows": true,
            },
          },
        ],
        "title": "Sample layout",
        "view": Object {
          "center": Array [
            1.5,
            45.4,
          ],
          "projection": Object {
            "name": "EPSG:3857",
          },
          "resolution": 1000,
          "rotation": 0,
        },
      },
      "id": "test-shared-view-1",
      "version": 2,
    },
  },
]
`;

exports[`M20230806 SharedViews should migrate 1`] = `
Array [
  Object {
    "key": "test-shared-view-1",
    "value": Object {
      "entry": Object {
        "id": "test-shared-view-1",
        "layers": Array [],
        "scale": Object {
          "x": 30,
          "y": 100,
        },
        "textFrames": Array [
          Object {
            "content": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Hey man, how are you ?",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "id": "test-text-frame-1",
            "position": Object {
              "x": 10,
              "y": 10,
            },
            "size": Object {
              "height": 100,
              "width": 100,
            },
            "style": Object {
              "background": "#FFFFFF",
              "withBorders": true,
              "withShadows": true,
            },
          },
        ],
        "title": "Sample layout",
        "view": Object {
          "center": Array [
            1.5,
            45.4,
          ],
          "projection": Object {
            "name": "EPSG:3857",
          },
          "resolution": 1000,
          "rotation": 0,
        },
      },
      "id": "test-shared-view-1",
      "version": 2,
    },
  },
]
`;

exports[`M20230806 Tiles should do nothing 1`] = `
Array [
  Object {
    "key": "http://1",
    "value": Object {
      "image": "blob data 1",
      "url": "http://1",
      "version": 99,
    },
  },
]
`;

exports[`M20230806 Tiles should migrate 1`] = `
Array [
  Object {
    "key": "http://1",
    "value": Object {
      "image": "blob data 1",
      "url": "http://1",
      "version": 2,
    },
  },
]
`;
